lib/internal/modules/esm/loader.js

```diff
@@ -202,33 +211,67 @@ class ESMLoader {
       const { ModuleWrap, callbackMap } = internalBinding('module_wrap');
       const module = new ModuleWrap(url, undefined, source, 0, 0);
       callbackMap.set(module, {
-        importModuleDynamically: (specifier, { url }) => {
-          return this.import(specifier, url);
+        importModuleDynamically: (specifier, { url }, import_assertions) => {
+          return this.import(specifier, url, import_assertions);
         }
       });
 
       return module;
     };
     const job = new ModuleJob(this, url, evalInstance, false, false);
-    this.moduleMap.set(url, job);
+    this.moduleMap.set(url, job, undefined);
     const { module } = await job.run();
 
     return {
       namespace: module.getNamespace(),
     };
   }
 
-  async getModuleJob(specifier, parentURL) {
+  async getModuleJob(specifier, parentURL, import_assertions) {
+    if (!ArrayPrototypeIncludes(supportedTypes, import_assertions.type)) {
+      throw new ERR_INVALID_IMPORT_ASSERTION('type', import_assertions.type);
+    }
     const { format, url } = await this.resolve(specifier, parentURL);
+
     let job = this.moduleMap.get(url);
-    // CommonJS will set functions for lazy job evaluation.
-    if (typeof job === 'function') this.moduleMap.set(url, job = job());
 
-    if (job !== undefined) return job;
+    if (job != null) {
+      const currentImportAssertionType = importAssertionTypeCache.get(job);
+      if (currentImportAssertionType === import_assertions.type) return job;
+
+      try {
+        // To avoid race conditions, wait for previous module to fulfill first.
+        await job.modulePromise;
```

Comment:
i don't think we can have any awaits between `moduleMap.get` and `moduleMap.set`, otherwise two imports could overlap during this await and both write to the cache below.

Reply:
Done in https://github.com/nodejs/node/pull/39921/commits/180099e74e6173e2cfc41d4abf3a4f53c68203bb.